app/controllers/concerns/two_factor_authenticatable_methods.rb

```diff
@@ -17,7 +17,15 @@ def handle_valid_verification_for_authentication_context(auth_method:)
 
     if IdentityConfig.store.feature_new_device_alert_aggregation_enabled && new_device?
       if current_user.sign_in_new_device_at.blank?
-        current_user.update(sign_in_new_device_at: disavowal_event.created_at)
+        timeframe_expired_at = current_user.events.where(
+          event_type: 'sign_in_notification_timeframe_expired',
+        ).order(created_at: :desc).limit(1).last
+        if timeframe_expired_at.present?
+          current_user.update(sign_in_new_device_at: timeframe_expired_at.created_at)
```

Comment:
1. This sort of querying could also be good to move into a method. A model or service class method would be best so that it could be tested in isolation, but I think it could be fine enough to just make it a private method of this concern. We'd want to memoize it to make sure we only evaluate it once.
2. The suffix "_at" implies to me the value is a date. But in this case it's an event. I think the name should reflect that.

```rb
def timeframe_expired_event
  return @timeframe_expired_event if defined?(@timeframe_expired_event)
  @timeframe_expired_event = current_user.events.where(
    event_type: 'sign_in_notification_timeframe_expired',
  ).order(created_at: :desc).limit(1).take
end
```